src/main/scala/gemmini/TilerScheduler.scala

```diff
@@ -0,0 +1,452 @@
+//===========================================================================
+// TilerController's Internal Scheduler implementation
+//===========================================================================
+package gemmini
+
+import chisel3._
+import chisel3.util._
+import freechips.rocketchip.config._
+import freechips.rocketchip.tile._
+import GemminiISA._
+import Util._
+
+class TilerScheduler[T <: Data: Arithmetic, U <: Data]
+  (config: GemminiArrayConfig[T,U])(implicit val p: Parameters) 
+  extends Module with HasCoreParameters {
+  import config._
+
+  //=========================================================================
+  // interface
+  //=========================================================================
+  val io = IO(new Bundle {
+    val cmd_in = Flipped(Decoupled(new RoCCCommand))
+    val issue = new Bundle {
+      val exec  = Decoupled(new GemminiCmd(ROB_ENTRIES))
+      val load  = Decoupled(new GemminiCmd(ROB_ENTRIES))
+      val store = Decoupled(new GemminiCmd(ROB_ENTRIES))
+    }
+    val completed = Flipped(Decoupled(UInt(LOG2_ROB_ENTRIES.W)))
+    val busy = Output(Bool())
+  })
+  io.completed.ready := true.B
+
+  //=========================================================================
+  // ...
+  //=========================================================================
+  val ldq :: stq :: exq :: Nil = Enum(3)
+  val q_t = ldq.cloneType
+
+  val MAX_CMD_ID   = 1000000            // FOR DEBUGGING ONLY
+  val debug_cycle  = RegInit(0.U(32.W)) // FOR DEBUGGING ONLY
+
+  debug_cycle := debug_cycle + 1.U      // FOR DEBUGGING ONLY
+  val cmd_id = Counter(MAX_CMD_ID)      // FOR DEBUGGING ONLY
+
+  // scratchpad range, used for tracking RAW, WAR, and WAW deps
+  class SPRange extends Bundle {
+    val valid  = Bool()
+    val is_acc = Bool()
+    val start  = UInt(30.W) // TODO magic number
+    val end    = UInt(30.W) // TODO magic number
+    def overlaps(other: SPRange) = valid && other.valid && 
+                                   (is_acc === other.is_acc) &&
+                                   (start < other.end) && 
+                                   (end > other.start)
+  }
+
+  class Entry extends Bundle {
+    val q = q_t.cloneType
+
+    val is_config = Bool()
+
+    val cmd_id     = UInt(log2Up(MAX_CMD_ID).W)   // FOR DEBUGGING ONLY
+    val is_load    = Bool()  // also true on config_load.  FOR DEBUGGING ONLY
+    val is_store   = Bool()  // also true on config_store. FOR DEBUGGING ONLY
+    val is_exec    = Bool()  // also true on config_ex.    FOR DEBUGGING ONLY
+    val is_preload = Bool()  // FOR DEBUGGING ONLY
+
+    val op1 = new SPRange()
+    val op2 = new SPRange()
+    val dst = new SPRange()
+
+    val issued = Bool()
+
+    val complete_on_issue = Bool()
+
+    val cmd = new RoCCCommand
+
+    val deps = Vec(ROB_ENTRIES, Bool())
+    def ready(dummy: Int = 0): Bool = !deps.reduce(_ || _)
+  }
+
+  val entries = Reg(Vec(ROB_ENTRIES, UDValid(new Entry)))
```

Comment:
Are we essentially duplicating the ROB here, so that we have one ROB for the RISC instructions, and another for the CISC tiler?

Reply:
pretty much. The TilerScheduler is probably about 99% the same functionality as the ROB. It would've been a bit more work to try to merge them. So to me, it would make sense to do that in a later PR if desired